rust/agama-network/src/dbus/tree.rs

```diff
@@ -0,0 +1,206 @@
+use agama_lib::error::ServiceError;
+use parking_lot::Mutex;
+use uuid::Uuid;
+
+use crate::{action::Action, dbus::interfaces, model::*};
+use std::collections::HashMap;
+use std::sync::mpsc::Sender;
+use std::sync::Arc;
+
+const CONNECTIONS_PATH: &str = "/org/opensuse/Agama/Network1/connections";
+const DEVICES_PATH: &str = "/org/opensuse/Agama/Network1/devices";
+
+/// Handle the objects in the D-Bus tree for the network state
+pub struct Tree {
+    connection: zbus::Connection,
+    actions: Sender<Action>,
+    objects: Arc<Mutex<ObjectsRegistry>>,
+}
+
+impl Tree {
+    /// Creates a new tree handler.
+    ///
+    /// * `connection`: D-Bus connection to use.
+    /// * `actions`: sending-half of a channel to send actions.
+    pub fn new(connection: zbus::Connection, actions: Sender<Action>) -> Self {
+        Self {
+            connection,
+            actions,
+            objects: Default::default(),
+        }
+    }
+
+    /// Adds devices to the D-Bus tree.
+    ///
+    /// * `devices`: list of devices.
+    pub async fn add_devices(&mut self, devices: &Vec<Device>) -> Result<(), ServiceError> {
+        for (i, dev) in devices.iter().enumerate() {
+            let path = format!("{}/{}", DEVICES_PATH, i);
+            self.add_interface(&path, interfaces::Device::new(dev.clone()))
+                .await?;
+            let mut objects = self.objects.lock();
+            objects.register_device(&dev.name, &path);
+        }
+
+        self.add_interface(
+            DEVICES_PATH,
+            interfaces::Devices::new(Arc::clone(&self.objects)),
+        )
+        .await?;
+
+        Ok(())
+    }
+
+    /// Adds connections to the D-Bus tree.
+    ///
+    /// * `connections`: list of connections.
+    pub async fn add_connections(&self, connections: &Vec<Connection>) -> Result<(), ServiceError> {
+        for conn in connections.iter() {
+            self.add_connection(conn).await?;
+        }
+
+        self.add_interface(
+            CONNECTIONS_PATH,
+            interfaces::Connections::new(Arc::clone(&self.objects), self.actions.clone()),
+        )
+        .await?;
+
+        Ok(())
+    }
+
+    /// Adds a connection to the D-Bus tree.
+    ///
+    /// * `connection`: connection to add.
+    pub async fn add_connection(&self, conn: &Connection) -> Result<(), ServiceError> {
+        let mut objects = self.objects.lock();
+
+        let path = format!("{}/{}", CONNECTIONS_PATH, objects.connections.len());
```

Comment:
this is interesting. So I failed to see renumbering  in remove_connection ( it can be of course my fault :wink: ), so what happen if there is add + add + removal first + add? Does the last add has same connection number as add2?

Reply:
Yes, you are fully right. I have changed the API to prevent such situations. Actually, as I mentioned in a TODO, we should improve the mechanism to "refresh" the list of connections and devices (avoiding the brute-force approach).